src/transformers/models/t5/convert_t5x_checkpoint_to_pytorch.py

```diff
@@ -0,0 +1,221 @@
+        new[f"decoder.block.{i}.layer.1.EncDecAttention.q.weight"] = q.T
+        new[f"decoder.block.{i}.layer.1.EncDecAttention.v.weight"] = v.T
+
+        # Block i, layer 2 (MLP).
+        layer_norm = t5x_layer_norm_lookup(old, i, "decoder", "pre_mlp_layer_norm")
+        wi, wo = t5x_mlp_lookup(old, i, "decoder", split_mlp_wi)
+        new[f"decoder.block.{i}.layer.2.layer_norm.weight"] = layer_norm
+        if split_mlp_wi:
+            new[f"decoder.block.{i}.layer.2.DenseReluDense.wi_0.weight"] = wi[0].T
+            new[f"decoder.block.{i}.layer.2.DenseReluDense.wi_1.weight"] = wi[1].T
+        else:
+            new[f"encoder.block.{i}.layer.2.DenseReluDense.wi.weight"] = wi.T
+        new[f"decoder.block.{i}.layer.2.DenseReluDense.wo.weight"] = wo.T
+
+    new["decoder.final_layer_norm.weight"] = old["decoder/decoder_norm/scale"]
+    new["decoder.block.0.layer.0.SelfAttention.relative_attention_bias.weight"] = old[
+        "decoder/relpos_bias/rel_embedding"
+    ].T
+
+    # LM Head (only in v1.1 checkpoints, in v1.0 embeddings are used instead)
+    if "decoder/logits_dense/kernel" in old:
+        new["lm_head.weight"] = old["decoder/logits_dense/kernel"].T
+
+    return new
+
+
+def make_state_dict(converted_params):
+    """Prepares a state dict for the PyTorch model."""
+    # Make a state dict with torch tensors.
+    state_dict = collections.OrderedDict([(k, torch.from_numpy(v.copy())) for (k, v) in converted_params.items()])
+
+    # Add what is missing.
+    if "encoder.embed_tokens.weight" not in state_dict:
+        state_dict["encoder.embed_tokens.weight"] = state_dict["shared.weight"]
+
+    if "decoder.embed_tokens.weight" not in state_dict:
+        state_dict["decoder.embed_tokens.weight"] = state_dict["shared.weight"]
+
+    if "lm_head.weight" not in state_dict:  # For old 1.0 models.
+        print("Using shared word embeddings as lm_head.")
+        state_dict["lm_head.weight"] = state_dict["shared.weight"]
+
+    return state_dict
+
+
+def load_t5x_weights_in_t5(model, config, t5x_checkpoint_path):
+    """Replaces the params in model witht the T5X converted params."""
+    num_weights = sum([1 for _ in model.named_parameters()])
+    num_layers = config.num_layers
+
+    variables = checkpoints.load_t5x_checkpoint(t5x_checkpoint_path)
+    converted = convert_t5x_to_pytorch(variables, num_layers=num_layers)
+    if len(converted) != num_weights:
+        raise RuntimeError(
+            f"Number of converted parameters {len(converted)} does not match "
+            f"number of original parameters {num_weights}"
+        )
+    state_dict = make_state_dict(converted)
+    model.load_state_dict(state_dict, strict=True)
+
+
+def convert_t5x_checkpoint_to_pytorch(t5x_checkpoint_path, config_file, pytorch_dump_path):
+    """Loads the config and model, converts the T5X checkpoint, and saves a PyTorch checkpoint."""
+    # Initialise PyTorch model
+    config = T5Config.from_json_file(config_file)
```

Comment:
I presume we're loading a Transformers-compatible JSON file here? It would be cool if we could get the config directly from the T5X checkpoint and load the Transformers `T5Config` with the correct values! https://github.com/google-research/t5x/blob/a1dd1d732112daa53f6a28766da79814abeb8144/t5x/examples/t5/local_tiny.gin#L32

Reply:
Yeah it would but converting `gin` to `json` is not generally possible. I think it's fair to put it on the user to modify a `json` config if needed, but in most cases right now the corresponding `json` is already available on HF.

Reply:
Sure! If the corresponding `json` config is already available in HF that works for me!